### PR TITLE
Add optional attribute to set the junit results parse interval

### LINKED ISSUE #21: Specify junit realtime result parse rate
_Originally reported by @dcendents in [JENKINS-50046](https://issues.jenkins.io/browse/JENKINS-50046)_

Currently the plugin will parse results every 1/100 of the estimated build time and never faster than every 5 seconds.

There is a todo to put a max value at 1 minute but it's not done yet.

I don't mind a default logic like that but I'd prefer if the plugin had another attribute available where I can manually specify the refresh rate I want.

 

The default logic does not work at all for me because the estimated build time is never consistent, we use the lock plugin to lock resources before running integration tests and the availability of those resources make the build time vary a lot from one build to another.

What could work though is using the estimated build time for the specific stage we are in (if that value exists), because this would be the actual running time without the waiting period trying to acquire the lock

### PR DESCRIPTION
**Pull request by dcendents**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Adding an optional attribute to set the junit results parse interval.
When the value is not set, the previous default logic is applied.

Fixes #21 

I did not add unit tests, this is hard to test, but here is some output to prove it works:

Configuration page:

image

Logging output when the parse interval is not set (never more often then every 5 seconds)

```
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.PipelineRealtimeTestResultAction
parsing *.xml in /var/jenkins_home/workspace/junitrealtime on node  for junitrealtime #5
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Parsing of 4 test results took 7ms
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:07 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:08 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:09 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:10 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:11 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 9:57:14 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.PipelineRealtimeTestResultAction
parsing *.xml in /var/jenkins_home/workspace/junitrealtime on node  for junitrealtime #5
Jan 03, 2022 9:57:14 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Parsing of 4 test results took 6ms

```

Logging output when the parse interval is set to 500 (milliseconds)

```
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.PipelineRealtimeTestResultAction
parsing *.xml in /var/jenkins_home/workspace/junitrealtime on node  for junitrealtime #6
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Parsing of 4 test results took 5ms
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:25 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.PipelineRealtimeTestResultAction
parsing *.xml in /var/jenkins_home/workspace/junitrealtime on node  for junitrealtime #6
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Parsing of 4 test results took 5ms
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Cache hit
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.PipelineRealtimeTestResultAction
parsing *.xml in /var/jenkins_home/workspace/junitrealtime on node  for junitrealtime #6
Jan 03, 2022 10:13:26 PM FINE org.jenkinsci.plugins.junitrealtimetestreporter.AbstractRealtimeTestResultAction
Parsing of 4 test results took 5ms

```

### CODE REVIEW

**Comment by timja** [comment]
conflicted

**Comment by dcendents** [comment]
@timja I have rebased the branch and tested it again

**Comment by timja** [comment]
Thanks I’ll try review in the next couple of days

File: src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java [hunk 1]
```diff
@@ -132,6 +133,17 @@ public void setSkipMarkingBuildUnstable(boolean skipMarkingBuildUnstable) {
         this.skipMarkingBuildUnstable = skipMarkingBuildUnstable;
     }
 
+    public Long getParseInterval() {
+        return parseInterval;
+    }
+
+    @DataBoundSetter
+    public void setParseInterval(Long parseInterval) {
+        if (parseInterval == null || parseInterval.longValue() > 0) {
```
Comment (timja) [inline src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java:142]: this doesn't look right, if the value is null then set is? it's only ever going to set null here?

File: src/main/resources/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep/help-parseInterval.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<div>
+    <p>
+        The interval between 2 file system scans to parse test results (in milliseconds).
```
Comment (timja) [inline src/main/resources/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep/help-parseInterval.html:3]:
```suggestion
        The interval between two file system scans to parse test results (in milliseconds).
```

File: src/main/resources/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep/help-parseInterval.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+        Must be a positive value greater than 0.
+        Default value is calculated as 1/100 of the estimated build duration, but never more often than every 5 seconds.
```
Comment (timja) [inline src/main/resources/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep/help-parseInterval.html:5]: milliseconds seems to be an odd unit for this field? would you ever actually want to do it that low?

File: src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java (hunk 1 above)
Comment (dcendents) [reply src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java:142]: We don't want a negative value (parse interval needs to be positive and higher than 0)
We allow null values (parse interval not set explicitely), then the current logic of "every 1/100 of a job estimated duration but not more often than every 5 seconds" will apply.

Actually I'm giving the option of setting a specific parse interval without changing the logic for existing builds. I personally find the current dynamic logic a bit weird, but unless we throw that to the bin and set a default value that is static, we need a way for the users to express they want the dynamic calculated parse rate, that is what the null value is.

But I'm open to suggestions if you think we can completely change the current behavior.

Comment (timja) [reply src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java]: I don’t understand your comment, the check was the wrong way round before from what I can tell but it seems fixed now :)

**Comment by timja** [comment]
I’ll do a test of this later on but looks good

**Comment by timja** [comment]
Pipeline syntax wasn't round tripping, it was x1000 of the value,

```
realtimeJUnit(parseInterval: 1000, testResults: 'abcd') {
    // some block
}
```

Fixed in https://github.com/jenkinsci/junit-realtime-test-reporter-plugin/pull/32/commits/5a550bdd715d298a8a9b4b4b7dfacc102b4c31a4

**Pull request merged**